### PR TITLE
feat(to_hbjson): Write hbjson from a Honeybee model

### PR DESCRIPTION
**Pull request by devngc**

### CODE REVIEW

**Comment by devngc** [comment]
@chriswmackey,
This is ready for review.
Thanks!

File: honeybee/model.py [hunk 1]
```diff
@@ -1349,6 +1351,46 @@ def to_dict(self, included_prop=None, triangulate_sub_faces=False):
...
+                Apertures and Doors) should be triangulated if they have more than
+                4 sides (True) or whether they should be left as they are (False).
+                This triangulation is necessary when exporting directly to EnergyPlus
+                since it cannot accept sub-faces with more than 4 vertices. Note that
+                setting this to True will only triangulate sub-faces with parent Faces
+                that also have parent Rooms since orphaned Apertures and Faces are
+                not relevant for energy simulation. Default: False.
+        """
+        # Create dictionary from the Honeybee Model
+        hb_dict = self.to_dict(included_prop=None, triangulate_sub_faces=False)
```
Comment (mostaphaRoudsari) [inline honeybee/model.py:1376]:
```suggestion
        hb_dict = self.to_dict(included_prop=included_prop, triangulate_sub_faces=triangulate_sub_faces)
```

File: honeybee/model.py [hunk 2]
```diff
@@ -1349,6 +1351,46 @@ def to_dict(self, included_prop=None, triangulate_sub_faces=False):
...
+
+        # Setting up a name for the HBJSON
+        if not name.lower().endswith('.hbjson'):
+            file_name = '{}.hbjson'.format(name)
+        else:
+            name = name[:-7]
```
Comment (mostaphaRoudsari) [inline honeybee/model.py:1382]: I'm not sure why this is needed here?

**Review by mostaphaRoudsari — COMMENTED** [review]
I added two comments while we are waiting for @chriswmackey!

File: honeybee/model.py (hunk 2 above)
Comment (devngc) [reply honeybee/model.py:1382]: This is in case a user gives a file name of ,"testhbjson".

Comment (mostaphaRoudsari) [reply honeybee/model.py:1383]: Hmm... why don't you just use `test.hbjson`? You are removing the extension in this line and then adding it back again in next line?

Comment (devngc) [reply honeybee/model.py:1383]: I get your point. I am removing `.hbjson` from the name and adding back in. Whereas I could just use that as a name. I think I will change that.

**Review by mostaphaRoudsari — APPROVED** [review]
LGTM! Thanks.

File: honeybee/model.py [hunk 3]
```diff
@@ -1349,6 +1351,46 @@ def to_dict(self, included_prop=None, triangulate_sub_faces=False):
 
         return base
 
+    def to_hbjson(self, name="unnamed", folder_path=None, indent=None, included_prop=None, triangulate_sub_faces=False):
```
Comment (chriswmackey) [inline honeybee/model.py:1354]: This is longer than 89 characters.

File: honeybee/model.py [hunk 4]
```diff
@@ -1349,6 +1351,46 @@ def to_dict(self, included_prop=None, triangulate_sub_faces=False):
...
+        """Writes a Honeybee model to HBJSON.
+
+        Args:
+            name: A text string that will be the name of the HBJSOn. Defaults to "unnamed".
```
Comment (chriswmackey) [inline honeybee/model.py:1358]: `n` of `HBJSON` should be uppercase. The same below.

File: honeybee/model.py [hunk 5]
```diff
@@ -1349,6 +1351,46 @@ def to_dict(self, included_prop=None, triangulate_sub_faces=False):
...
+            folder_path: A text string of path to folder where HBJSOn will be written. Defaults to None.
```
Comment (chriswmackey) [inline honeybee/model.py:1359]: You should explain what the `None` means. Same below.

File: honeybee/model.py [hunk 6]
```diff
@@ -1349,6 +1351,46 @@ def to_dict(self, included_prop=None, triangulate_sub_faces=False):
...
+        """
+        # Create dictionary from the Honeybee Model
+        hb_dict = self.to_dict(included_prop=included_prop,
+                               triangulate_sub_faces=triangulate_sub_faces)
+
+        # Setting up a name for the HBJSON
+        if name.lower().endswith('.hbjson'):
+            file_name = name
+        else:
+            file_name = '{}.hbjson'.format(name)
```
Comment (chriswmackey) [inline honeybee/model.py:1383]: You can do this more elegantly with a couple of lines:
```
file_name  = name if name.lower().endswith('.hbjson') or name.lower().endswith('.json') \
    else '{}.hbjson'.format(name)
```

File: honeybee/model.py (hunk 3 above)
Comment (chriswmackey) [reply honeybee/model.py:1354]: Also, can the default indent just be 0?

**Review by chriswmackey — COMMENTED** [review]
Generally looks good. I just had a few comments.

**Comment by chriswmackey** [comment]
And one more comment:
We could really use a test for this code.  Maybe you can include it in the existing `Model.to_dict()` method.

File: honeybee/model.py (hunk 3 above)
Comment (devngc) [reply honeybee/model.py]: Revised

File: honeybee/model.py (hunk 4 above)
Comment (devngc) [reply honeybee/model.py]: Revised

File: honeybee/model.py (hunk 5 above)
Comment (devngc) [reply honeybee/model.py]: Explanation added

File: honeybee/model.py (hunk 6 above)
Comment (devngc) [reply honeybee/model.py]: Revised

**Comment by devngc** [comment]
@chriswmackey, A new test is added in `model_test` module for the the `to_hbjson` method

File: honeybee/model.py [hunk 7]
```diff
@@ -1349,6 +1351,50 @@ def to_dict(self, included_prop=None, triangulate_sub_faces=False):
...
+                  included_prop=None, triangulate_sub_faces=False):
+        """Writes a Honeybee model to HBJSON.
+
+        Args:
+            name: A text string that will be the name of the HBJSON.
+                Defaults to "unnamed" for the file name for HBJSON.
+            folder_path: A text string of path to folder where HBJSON will be written.
+                Defaults to None. If folder_path is not specified, the default simulation
+                folder will be used to write the HBJSON. This default simulation folder
+                is at "C:\\Users\\USERNAME\\simulation."
```
Comment (chriswmackey) [inline honeybee/model.py:1364]: This is alright but, for future reference, this is only for Windows machines.  It's different on Mac and Linux.  So it's probably better not to state it in the docstring.

**Review by chriswmackey — APPROVED** [review]
Looks good. Let's merge.

File: honeybee/model.py (hunk 7 above)
Comment (devngc) [reply honeybee/model.py:1364]: Noted

**Pull request merged**